vscode-utils/vscodeEnv: add vscodeWithConfiguration, vscodeExts2nix a…

countoren opened the pull request:
based on this [project](https://github.com/countoren/VSCodeEnv) 

#### Motivation for this change

Declarative configuration of vscode settings and extensions in a per project basis.

And I was trying to solve these problems:

- some extensions mutate the contents of their folder. Therefore, the nix store is not the best place for those extensions.

- being able to automatically get the extensions' information that nix needs to fetch them. Also, use vscode interface to manage the extensions while having a nix expression defining the extensions that are used for a project.

- VSCode Global configuration folder and globally enabled extensions.
 

#### Things done
created :
- vsCodeWithConfiguration - wrapper for vscode that:
  - copy extensions to local extensions dir. mutable extensions will get copied while nix extensions will get symlinked.
  - set user-data folder to the template path of the nix-shell in order to avoid global settings(those could be configured in nix) and symlink global settings.json and keybindings.json to the local one in the project's .vscode folder.

- updateSettings - shell script that creates or overrides all settings files in .vscode based on the nix expression. 
settings.json will be initialized with :   
```json 
{  
   "extensions.autoCheckUpdates": false,
  "extensions.autoUpdate": false,
  "update.mode": "none"
}
```

- vscodeExtensions2nix - with a vscode given as parameter, it will run with "--list-extensions" to output a valid nix expression that returns a list of sets of this expression (including sha).

- vscodeEnv - puts together vscodeWithConfiguration, vscodeExtensions2nix and updateSettings. Creates another wrapper above vscodeWithConfiguration which adds "--wait" flag in order to run vscodeExtensions2nix after vscode exits. vscodeExtensions2nix will update an expression file with the mutable extensions currently installed.

#### Notify maintainers

cc @zimbatm

countoren committed:
vscode-utils/vscodeEnv: add vscodeWithConfiguration, vscodeExts2nix and vscodeEnv

move mktplcExtRefToFetchArgs to file in order to be shared with the new derivations(privately)

countoren committed:
vscode-utils/vscodeEnv: add vscodeWithConfiguration, vscodeExts2nix and vscodeEnv

move mktplcExtRefToFetchArgs to file in order to be shared with the new derivations(privately)

zimbatm commented on pkgs/misc/vscode-extensions/vscode-utils.nix:
```diff
@@ -80,10 +73,24 @@ let
-in
+  vscodeWithConfiguration = (userParams : import ./vscodeWithConfiguration.nix {
+   inherit lib vscode extensionsFromVscodeMarketplace writeShellScriptBin;
+  } // userParams);
```
the parenthesis are not needed here

countoren replied:
fix in latest commit

zimbatm commented on pkgs/misc/vscode-extensions/vscode-utils.nix:
```diff
@@ -80,10 +73,24 @@ let
+  
+  vscodeExts2nix = (userParams : import ./vscodeExts2nix.nix {
+    inherit lib vscode;
+  } // userParams);
```
same

zimbatm commented on pkgs/misc/vscode-extensions/vscode-utils.nix:
```diff
@@ -80,10 +73,24 @@ let
+
+  vscodeEnv = (userParams : import ./vscodeEnv.nix {
+    inherit lib writeShellScriptBin extensionsFromVscodeMarketplace vscode;
+  } // userParams );
```
same

countoren replied:
fix in latest commit

zimbatm commented on pkgs/misc/vscode-extensions/vscodeEnv.nix:
```diff
@@ -0,0 +1,40 @@
+  };
+  code = writeShellScriptBin "code" ''
+    ${vscodeWithConfiguration}/bin/code --wait "$@" 
+    echo 'running vscodeExts2nix to update ${mutableExtensionsFilePath}...'
```
```suggestion
    echo 'running vscodeExts2nix to update ${toString mutableExtensionsFilePath}...'
```

zimbatm commented on pkgs/misc/vscode-extensions/vscodeEnv.nix:
```diff
@@ -0,0 +1,40 @@
+  code = writeShellScriptBin "code" ''
+    ${vscodeWithConfiguration}/bin/code --wait "$@" 
+    echo 'running vscodeExts2nix to update ${mutableExtensionsFilePath}...'
+    ${vscodeExts2nix}/bin/vscodeExts2nix > ${mutableExtensionsFilePath}
```
```suggestion
    ${vscodeExts2nix}/bin/vscodeExts2nix > ${toString mutableExtensionsFilePath}
```
this is to avoid having the path be added to the /nix/store

zimbatm commented on pkgs/misc/vscode-extensions/vscodeEnv.nix:
```diff
@@ -0,0 +1,40 @@
+, vscode                          ? pkgs.vscode
+}:
+let  
+  mutableExtensionsFilePath = builtins.toPath mutableExtensionsFile;
```
this is not needed

zimbatm commented on pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix:
```diff
@@ -0,0 +1,39 @@
+{ pkgs                            ? import <nixpkgs> {}
+, lib                             ? pkgs.lib
+, writeShellScriptBin             ? pkgs.writeShellScriptBin
+, extensionsFromVscodeMarketplace ? pkgs.vscode-utils.extensionsFromVscodeMarketplace
```
```suggestion
, extensionsFromVscodeMarketplace ? pkgs.vscode-utils.extensionsFromVscodeMarketplace
, vscode
```

zimbatm commented on pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix:
```diff
@@ -0,0 +1,39 @@
+
+##User input
+
+, vscode                          ? pkgs.vscode
```
```suggestion
, vscode                          ? orig.vscode
```

countoren committed:
vsode-utils: extracting attributes to limit input range

countoren committed:
vsode-utils: extracting attributes to limit input range

countoren committed:
vscode-utils/vscodeEnv: split to 2 functions vscodeWithConfiguration, vscodeExts2nix, vscodeEnv

change usage of toPath with toString

countoren committed:
vscode-utils/vscodeEnv: split to 2 functions vscodeWithConfiguration, vscodeExts2nix, vscodeEnv

change usage of toPath with toString

countoren commented:
@zimbatm updated base on your comments.

strager commented on pkgs/misc/vscode-extensions/vscodeEnv.nix:
```diff
@@ -0,0 +1,45 @@
+#use vscodeWithConfiguration and vscodeExts2nix to create vscode exetuable that when exits(vscode) will update the mutable extension file, which is imported when getting evaluated by nix.
```
Fix typo: s/exetuable/executable/

> #use vscodeWithConfiguration and vscodeExts2nix to create vscode executable that when exits(vscode) will update the mutable extension file, which is imported when getting evaluated by nix.

I'd also improve some of the grammar as follows:

> #Use vscodeWithConfiguration and vscodeExts2nix to create a vscode executable. When the executable exits, it updates the mutable extension file, which is imported when evaluated by Nix later.

countoren replied:
updated

strager commented on pkgs/misc/vscode-extensions/vscodeWithConfiguration.nix:
```diff
@@ -0,0 +1,38 @@
+  cpExtensions = ''
+    ${lib.concatMapStringsSep "\n" (e : ''ln -sfn ${e}/share/vscode/extensions/* ${vscodeExtsFolderName}/'') nixExtsDrvs}
+    ${lib.concatMapStringsSep "\n" (e : ''
+      cp -a ${e}/share/vscode/extensions/${e.vscodeExtUniqueId} ${vscodeExtsFolderName}/${e.vscodeExtUniqueId}-${(lib.findSingle (ext: ''${ext.publisher}.${ext.name}'' == e.vscodeExtUniqueId) "" "m" mutableExtensions ).version}
```
Which extensions have problems when their directory is immutable? Could we fix these extensions to avoid this copying?

countoren replied:
using this expression:
```
{ pkgs ? import <nixpkgs>{}}:
pkgs.vscode-with-extensions.override ({
    vscodeExtensions = pkgs.vscode-utils.extensionsFromVscodeMarketplace ([
    {
      name = "csharp";
      publisher = "ms-vscode";
      version = "1.21.5";
      sha256 = "031yx2kpjrw1b86y5nq110k9ywjgl26730ljddly05d8fxxham05";
    } 
    ]);
  })

will cuase this error in vscode:
```
Failed at stage: touchBeginFile
Error: EACCES: permission denied, mkdir '/nix/store/vnz8ja4qzmg4qpbr5am647djadclswsm-vscode-extensions-1.40.1/share/vscode/extensions/ms-vscode.csharp/.omnisharp'
```

and this error when trying to use nix-collect-garbage :
```
deleting '/nix/store/trash'
12428 store paths deleted, 664.94 MiB freed
error: cannot unlink '/nix/store/trash/4mnbcnp5fb8495pjlzwrfy24c9yk122n-vscode/extension-ms-vscode-csharp-1.21.5/share/vscode/extensions/ms-vscode.csharp/.omnisharp/1.34.5/license.md': Permission denied
```
im using darwin.

countoren committed:
vscode-utils/vscodeEnv: fix typo and grammer in the description comment

countoren committed:
vscode-utils/vscodeEnv: fix typo and grammer in the description comment

countoren commented:
@jonringer, what do you think about this PR?

countoren committed:
add updateSettings drv which will union nix settings configurations into the a vscode settings file

countoren committed:
use updateSettings in vscodeEnv in order to create and/or update settings.json, keybindings.json

countoren committed:
add missing jq to vscodeEnv

countoren committed:
change vscodeSettingsFile parameter type from path to string

countoren committed:
vscodeEnv updateSettings for keybindings change keybindings file name parameter type from path to string

countoren committed:
fix permission problem of mutableExtensions

peti commented:
@zimbatm, are you still interested in this PR? It's been sitting in the queue for quite a while now and it does look pretty useful to me. IMHO, the submitter has addressed your comments so I wonder whether there's a path forward for this PR to get merged?

zimbatm commented:
@peti: it seems like a good idea and @countoren put quite a bit of effort in it. It's a mix between the pure and impure universes which is quite a new approach.

The current complexity of the code is what prevented me from properly evaluating the PR. It would require me to block 1-2h to really look into it and see if it could be simplified. Which is what is slowing this PR down really.

countoren commented:
@peti , @zimbatm  thanks guys for looking at it :).
while this PR was waiting to the evaluation I copied the code as a new repo (https://github.com/countoren/VSCodeEnv) and updated a bit more one of the main things was to be able to "ignore" global configuration of VSCode (this is to have more guaranties of the configurations)  by moving vscode data folder to the tmp folder (instead of user folder) and symlink the user configuration files there to the project's one(which are overwritten from nix).
I am not sure if I should bring this feature before we merge or should we just look at this version first.

Another thing that I would like to solve here is the sudo that is done in order to keep the .vscode-exts folder in sync by removing non-nix-declared extensions, 
I'm not 100% sure it is needed because the folder should be created by the user that runs the wrapper (code executable) but I had some problem with it so I added it but it could be due to something else I was missing.

moinessim committed:
Use a different vscode user-data-dir for every project.
Treat workspace setting files as global for that user-data-dir with symlink.
Add updateLaunchCmd to update .vscode/launch.json.

countoren committed:
Merge branch 'updateSettings' into vscode-utils/vscodeEnv

moinessim commented:
This is a use example:
```nix
    let  
        personal-extensions-file = ./personal-extensions.nix;
    in
    vscodeEnv {
        nixExtensions = import ./extensions.nix 
          ++ (
                    if builtins.pathExists personal-extensions-file 
                    then import personal-extensions-file else []
                );
        mutableExtensionsFile = ./mutable-extensions.nix; 
        settings = 
        {
          "FSharp.useSdkScripts" = true;
        };
      }
```

moinessim committed:
vscode-extensions: Find mutable extensions by version in order not to delete them
Do not use sudo.

moinessim committed:
Merge branch 'vscode-utils/vscodeEnv' of https://github.com/countoren/nixpkgs-1 into vscode-utils/vscodeEnv

* 'vscode-utils/vscodeEnv' of https://github.com/countoren/nixpkgs-1: (178391 commits)
  Use a different vscode user-data-dir for every project. Treat workspace setting files as global for that user-data-dir with symlink. Add updateLaunchCmd to update .vscode/launch.json.
  fix permission problem of mutableExtensions
  vscodeEnv updateSettings for keybindings change keybindings file name parameter type from path to string
  change vscodeSettingsFile parameter type from path to string
  add missing jq to vscodeEnv
  use updateSettings in vscodeEnv in order to create and/or update settings.json, keybindings.json
  add updateSettings drv which will union nix settings configurations into the a vscode settings file
  vscode-utils/vscodeEnv: fix typo and grammer in the description comment
  vscode-utils/vscodeEnv: split to 2 functions vscodeWithConfiguration, vscodeExts2nix, vscodeEnv
  vsode-utils: extracting attributes to limit input range
  vscode-utils/vscodeEnv: add vscodeWithConfiguration, vscodeExts2nix and vscodeEnv
  xfce.xfce4-icon-theme: update gtk dependency and add license
  xfce.xfce4-hardware-monitor-plugin: update meta
  xfce.xfce4-embed-plugin: update meta and buildInputs
  xfce.xfce4-mailwatch-plugin: remove broken status
  xfce.exo: add support for gtk2, besides gtk3
  xfce.xfdesktop: 4.14.1 -> 4.14.2
  xfce.xfce4-weather-plugin: 0.8.10 -> 0.10.1
  xfce.xfdashboard: 0.7.5 -> 0.7.7
  xfce.xfce4-timer-plugin: 1.6.0 -> 1.7.0
  ...

peti commented:
I believe this PR has been waiting long enough and I'm not aware of any controversy around it, i.e. no-one seems to have anything against merging it. So ... I'm merging it. :-)
